File: 3.longest-substring-without-repeating-characters.ts
/*
 * @lc app=leetcode id=3 lang=typescript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
// @ts-ignore
function lengthOfLongestSubstring(s: string): number {
    const str = s.split('');
    const minLens: number[] = new Array(s.length).fill(0);
    var maxMinLen = 0;
    for (var i = s.length - 1; i >= 0; i--) {
        const char = str[i];
        const substr = str.slice(i + 1);
        const lengthBetweenRepeats = substr.findIndex(c => c === char);
        const minLen = lengthBetweenRepeats >= 0 ? lengthBetweenRepeats + 1 : substr.length + 1;
        const sublist = [minLen, ...minLens.slice(i + 1, i + minLen)];
        minLens[i] = (sublist.length > 0) ? Math.min(...sublist.map((len, j) => len + j)) : 0;
        if (minLens[i] > maxMinLen) maxMinLen = minLens[i];
    }
    return maxMinLen;
};
// @lc code=end

File: 3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen_chars = set()
        left = 0
        longest_substring = 0
        for right in range(len(s)):
            while s[right] in seen_chars:
                seen_chars.remove(s[left])
                left += 1
            seen_chars.add(s[right])
            longest_substring = max(longest_substring, len(seen_chars))
        return longest_substring

# @lc code=end

